### PR TITLE
Fix comparison of constant with short float NAN values

### PR DESCRIPTION
**Pull request by blesniewski**
Issue: both compared constants had NAN values, if we perform a bitwise check before the normal float comparison the result is as it should 

47732

### CODE REVIEW

File: inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp [hunk 1]
```diff
@@ -339,13 +359,20 @@ struct Equal<double> {
 };
 
 template <>
```
Comment (pelszkow) [inline inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp:361]: Now I think all vectors might be merged in one:
```
template <typename T>
struct Equal<std::vector<T>> {
    static bool equal_value(const std::vector<T>& lhs, const std::vector<T>& rhs) {
        return lhs.size() == rhs.size() &&
               std::equal(begin(lhs), end(lhs), begin(rhs), Equal<T>::equal_value);
    }
};

```

File: inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp [hunk 2]
```diff
@@ -742,15 +777,51 @@ FunctionsComparator::Result FunctionsComparator::compare(
...
-                const auto equal = [](std::shared_ptr<Constant> c1, std::shared_ptr<Constant> c2) {
-                    const auto& c1v = c1->cast_vector<double>();
-                    const auto& c2v = c2->cast_vector<double>();
-
-                    return c1v.size() == c2v.size() && std::equal(
-                                                           begin(c1v), end(c1v), begin(c2v),
-                                                           [](const double& s1, const double& s2) {
-                                                               return std::abs(s1 - s2) < 0.001;
-                                                           });
+                const auto equal = [](std::shared_ptr<Constant> c1,
```
Comment (pelszkow) [inline inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp:780]: Move this lambda outside this function.

File: inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp [hunk 3]
```diff
@@ -742,15 +777,51 @@ FunctionsComparator::Result FunctionsComparator::compare(
...
+                        auto c2v = c2->cast_vector<ngraph::bfloat16>();
+                        rc = c1v.size() == c2v.size()
+                            ? Equal<std::vector<ngraph::bfloat16>>::equal_value(c1v, c2v)
+                            : false;
+                        break;
+                    }
+                    case ngraph::element::Type_t::f16: {
+                        const auto &c1v = c1->cast_vector<ngraph::float16>();
+                        const auto &c2v = c2->cast_vector<ngraph::float16>();
+                        rc = c1v.size() == c2v.size()
```
Comment (pelszkow) [inline inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp:802]: Don't see reason to use elvis operator.
This should works:
```
return Equal<std::vector<float16>>::equal_value(c1v, c2v);
```
1) `Equal` for `vector` check size
2) `rc` is not need
3) `c1v.size() == c2v.size() || Equal<vec>(c1v, c2v)` looks better.

File: inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp [hunk 4]
```diff
@@ -742,18 +777,51 @@ FunctionsComparator::Result FunctionsComparator::compare(
...
+                      rc = c1v.size() == c2v.size() &&
+                           Equal<std::vector<double>>::equal_value(c1v, c2v);
+                      break;
+                    }
+                    }
+                    return rc;
                 };
 
                 if (const1 && const2 && !equal(const1, const2)) {
+                    equal(const1, const2);
```
[the remark this answers is not in the document]
Comment (blesniewski) [reply inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp:824]: Sorry, left it there so I have easier job debugging it, forgot to remove it :)

**Comment by jdanieck** [comment]
@blesniewski is it possible to create simple unit test to prove that fix is working ?

File: inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp [hunk 5]
```diff
@@ -474,6 +451,45 @@ struct Equal<SpecialBodyPorts> {
     }
 };
 
+using Constant = ngraph::opset1::Constant;
+template <> struct Equal<std::shared_ptr<Constant>> {
+    static bool equal_value(const std::shared_ptr<Constant> lhs,
```
Comment (pelszkow) [inline inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp:456]: Copy of `shared_ptr` is required?

**Review by pelszkow — APPROVED** [review]

**Comment by blesniewski** [comment]
> @blesniewski is it possible to create simple unit test to prove that fix is working ?

Sure

File: inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp (hunk 5 above)
Comment (blesniewski) [reply inference-engine/tests/ie_test_utils/common_test_utils/ngraph_test_utils.cpp:456]: Guess not, I'll remove it

File: inference-engine/tests/functional/inference_engine/ir_serialization/models/add_abc_initializers_nan_const.prototxt [hunk 6]
```diff
@@ -0,0 +1,95 @@
+ir_version: 3
```
Comment (jdanieck) [inline inference-engine/tests/functional/inference_engine/ir_serialization/models/add_abc_initializers_nan_const.prototxt:1]: Looks like you missie adding this .prototxt to INSTANTIATE_TEST_CASE_P(ONNXSerialization)

**Review by jdanieck — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Pull request merged**